tt_metal/tt_metal.cpp

```diff
@@ -364,14 +364,27 @@ void CloseDevices(std::map<chip_id_t, Device *> devices) {
             for (uint32_t ts = t.size() - 1; ts > 0; ts--) {
                 if (devices.find(t[ts]) != devices.end()) {
                     devices[t[ts]]->close();
+                    // When a device is closed, its worker thread is joined. Stop tracking this
+                    // worker thread.
+                    tt::DevicePool::instance().unregister_worker_thread_for_device(devices[t[ts]]);
```

Comment:
please review this part, as discussed

Reply:
Thanks for pointing out the potential race here. I reviewed the code and found that we synchronize all worker threads when `CloseDevices` is called, before unregistering threads. This is the piece of code responsible for the synchronization, as soon as the function is called:

 ```
for (const auto &[device_id, dev] : devices) {
        dev->synchronize(); // Synchronize worker queue
        Synchronize(dev); // Synchronize device
}
``` 

Given that workers and the devices are idle, I believe that modifying the shared data structures should be safe.